packages/augur-state/src/db/BlockstreamListener.ts

```diff
@@ -0,0 +1,111 @@
+import {ExtendedLog} from "../api/types";
+import { Augur, ParsedLog, Log } from "@augurproject/api";
+import {Block, BlockAndLogStreamer, FilterOptions} from "ethereumjs-blockstream";
+import {ethers} from "ethers";
+import * as _ from "lodash";
+import {Provider} from "@augurproject/api";
+import {EthersProvider} from "ethers-provider";
+
+interface BlockstreamListenerDeps {
+  blockAndLogStreamer:BlockAndLogStreamer<Block, ExtendedLog>;
+  // TODO Use an emitter?
+  listenForNewBlocks: (callback: (block:Block) => void) => void;
+  parseLogs(logs: Log[]): ParsedLog[];
+}
+
+type LogCallbackType<T> = (blockHash: string, logs: T[]) => void;
+
+export class BlockstreamListener {
+  private logCallBacks:LogCallbackType<Log>[] = [];
+
+  constructor(private deps:BlockstreamListenerDeps) {
+    deps.blockAndLogStreamer.subscribeToOnLogsAdded(this.onLogsAdded);
+  }
+
+  public static create(provider:EthersProvider, parseLogs:(logs: Log[])=> ParsedLog[]) {
+    const getBlockByHashOrTag = (provider: EthersProvider) => async (hashOrTag: string | number): Promise<Block> => {
+      const block = await provider.getBlock(hashOrTag, false);
+      return {
+        number: "0x" + block.number.toString(16),
+        hash: block.hash,
+        parentHash: block.parentHash,
+      }
+    };
+
+    const getLogs = (provider: Provider) => async (filterOptions: FilterOptions): Promise<ExtendedLog[]> => {
+      const logs = await provider.getLogs({
+        ...filterOptions,
+        topics: _.compact(filterOptions.topics)
+      });
+
+      return logs.map((log) => ({
+        ...log,
+        logIndex: (log.logIndex || "0").toString(),
+        blockNumber: (log.blockNumber || "0").toString(),
+        blockHash: log.blockHash || "0",
```

Comment:
when do these fallback values actually occur?

Reply:
It makes typescript happy.